.github/github_workflow_scripts/purge_branch_protection_rules.py

```diff
@@ -0,0 +1,358 @@
+        logger.info(f"{rule} not deleted because it's in the list of protected rules '{','.join(PROTECTED_RULES)}'")
+    elif rule.matching_refs > 0:
+        logger.info(f"{rule} not deleted because it's associated to {rule.matching_refs} existing branches/refs")
+    else:
+        should = True
+
+    return should
+
+
+def write_deleted_summary_to_file(deleted: list[BranchProtectionRule]) -> None:
+    """
+    Helper function to create a Markdown summary file for deleted branches
+    if the `GITHUB_STEP_SUMMARY` is set. See
+    https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-job-summary
+
+    Arguments:
+    - `deleted` (``list[BranchProtectionRule]``): A list of deleted `BranchProtectionRule`s.
+    """
+
+    if os.getenv(GH_JOB_SUMMARY_ENV_VAR):
+        fp = Path(str(os.getenv(GH_JOB_SUMMARY_ENV_VAR)))
+
+        if not fp.is_file():
+            logger.warning(f"The env var {GH_JOB_SUMMARY_ENV_VAR} is not set to a file. Not writing summary...")
+            return
+
+        header = "## Deleted Branch Protection Rules"
+
+        if not deleted:
+            body = "### No branch protection rules were deleted"
+            markdown_content = f"{header}\n\n{body}\n"
+        else:
+            table_header = "| ID | Pattern | Matching Refs |\n| --- | ------- | ------------- |"
+            table_rows = [f"| {rule.id} | {rule.pattern} | {rule.matching_refs} |" for rule in deleted]
+            table_body = "\n".join(table_rows)
+            markdown_content = f"{header}\n\n{table_header}\n{table_body}\n"
+
+        logger.debug(f"Writing deleted jobs summary to Markdown to file '{fp}'...")
+        logger.debug(markdown_content)
+        fp.write_text(markdown_content)
+        logger.debug("Finished writing jobs summary to Markdown to file")
+    else:
+        logger.info(
+            f"Environmental variable '{GH_JOB_SUMMARY_ENV_VAR}' not set. Skipping writing job summary for deleted rules...")
+
+
+def get_token():
+    """
+    Helper method to retrieve the GitHub token
+    from the environmental variables.
+
+    Returns:
+    - `token` (``str``): The GitHub token.
+
+    Raises:
+    `OSError` if the `GITHUB_TOKEN` env var is not set.
+    """
+    token = os.getenv(GH_TOKEN_ENV_VAR)
+    if not token:
+        raise OSError(f"Error: The '{GH_TOKEN_ENV_VAR}' environment variable is not set.")
+    return token
+
+
+def send_request(gh_requester: github.Requester.Requester, query: str, variables: dict[str, str]) -> dict[str, str]:
+    """
+    Wrapper function to send a request to the GraphQL endpoint.
+
+    Arguments:
+    - `query` (``str``): The query to send.
+    - `variables` (``dict[str, str]``): The variables to send the query.
+
+    Returns:
+    - A `dict[str, str]` with the response.
+
+    Raises:
+    If the request fails, we raise
+    """
+
+    logger.debug("Sending GraphQL request...")
+    logger.debug(f"{query=}")
+    logger.debug(f"{variables=}")
```

Comment:
why not single log line?

Reply:
Since `query` (and possibly `variables`) are multiline, printing all of them in a single line could look a bit messy. I guess what I can do is have one `logger.debug` and add `\n` between each, e.g.:

```python
logger.debug("Sending GraphQL request...\n{query=}\n{variables})
```

WDYT?

Reply:
i'm ok with it